### ISSUE TITLE
Unused variable warning

### ISSUE THREAD

**Issue by preethint** [okbob/plpgsql_check#138]
Hi,
Found an observation while checking plpgsql_check_function. Below is the procedure that I am using. 
Here, variable **a** and **d** are used in an operation. Yes, they are of different data types.
But, plpgsql_check_function is throwing a warning that variable **a** and **d** are never used. Below is the code snippet for reference.
```
edb@40112=#create or replace procedure p1()
edb-# language plpgsql
edb-# as $$
edb$# declare
edb$#     a int;
edb$#     b int;
edb$#     c int;
edb$#     d char;
edb$# BEGIN
edb$#     c=a+d;
edb$# END; $$;
CREATE PROCEDURE

edb@40112=#select * from plpgsql_check_function('p1()', fatal_errors := false);
                                         plpgsql_check_function
---------------------------------------------------------------------------------------------------------
 error:42883:8:assignment:operator does not exist: integer + character
 Query: c=a+d
 --        ^
 Hint: No operator matches the given name and argument types. You might need to add explicit type casts.
 Context: at assignment to variable "c" declared on line 5
 warning:00000:3:DECLARE:unused variable "a"
 warning:00000:4:DECLARE:unused variable "b"
 warning:00000:6:DECLARE:unused variable "d"
 warning extra:00000:5:DECLARE:never read variable "c"
(9 rows)
```
Is this is an expected behaviour? If not, can this be improved?

**Comment by okbob** [comment]
Hi


po 6. 3. 2023 v 14:27 odesílatel preethint ***@***.***>
napsal:

> Hi,
> Found an observation while checking plpgsql_check_function. Below is the
> procedure that I am using.
> Here, variable *a* and *d* are used in an operation. Yes, they are of
> different data types.
> But, plpgsql_check_function is throwing a warning that variable *a* and
> *d* are never used. Below is the code snippet for reference.
>
> ***@***.***=#create or replace procedure p1()
> edb-# language plpgsql
> edb-# as $$
> edb$# declare
> edb$#     a int;
> edb$#     b int;
> edb$#     c int;
> edb$#     d char;
> edb$# BEGIN
> edb$#     c=a+d;
> edb$# END; $$;
> CREATE PROCEDURE
>
> ***@***.***=#select * from plpgsql_check_function('p1()', fatal_errors := false);
>                                          plpgsql_check_function
> ---------------------------------------------------------------------------------------------------------
>  error:42883:8:assignment:operator does not exist: integer + character
>  Query: c=a+d
>  --        ^
>  Hint: No operator matches the given name and argument types. You might need to add explicit type casts.
>  Context: at assignment to variable "c" declared on line 5
>  warning:00000:3:DECLARE:unused variable "a"
>  warning:00000:4:DECLARE:unused variable "b"
>  warning:00000:6:DECLARE:unused variable "d"
>  warning extra:00000:5:DECLARE:never read variable "c"
> (9 rows)
>
> Is this is an expected behaviour? If not, can this be improved?
>
Unfortunately, this is expected behaviour.

Probably it can be improved - but a) with a lot of redundant code from
pl_comp.c (because necessary code is not visible) or b) with a more than
less dirty second level of wrapping of parser hooks.
More the (b) can break this warning and in some situations can block true
warning. So the fix can be interesting, but not easy. Probably not too much
code, but with a lot of very dark blood magic. I'll try to fix it, but I
don't promise.

**Comment by kirkw** [comment]
Honestly, if fixing the syntax problem of a+d with ascii(d)

Makes the unused variable warning go away.  I am not sure it is worth fixing.  This is a typical "compiler/linter" issue, IMO, with regards to cascading error messages.


```
20:28:28 devuser@test= > create or replace procedure test.p1()
language plpgsql
as $$
declare
    a int;
    b int;
    c int;
    d char;
BEGIN
    c=a+ascii(d);
END; $$;
Time: 1.453 ms
20:28:38 devuser@test= > select * from plpgsql_check_function('test.p1()', fatal_errors := false);
                plpgsql_check_function
-------------------------------------------------------
 warning:00000:4:DECLARE:unused variable "b"
 warning extra:00000:5:DECLARE:never read variable "c"
(2 rows)

```

**Comment by surajkharage19** [comment]
@kirkw 
IMO, Fixing the plpgsql code would not be a good idea. Also, in this particular case, we are also expecting the error (operator does not exist). 
I agree with Pavel that fixing this would be challenging.

**Comment by okbob** [comment]
út 7. 3. 2023 v 3:38 odesílatel surajkharage19 ***@***.***>
napsal:

> @kirkw <https://github.com/kirkw>
> IMO, Fixing the plpgsql code would not be a good idea. Also, in this
> particular case, we are also expecting the error (operator does not exist).
> I agree with Pavel that fixing this would be challenging.
>

It should be fixed only on the plpgsql_check side.  But it is hard because
one possibility is wrapping up calling plpgsql from SQL parser, which is
more than dirty.

plpgsql_check is not a real compiler - it almost calls plpgsql parser's
routines without execution (plpgsql calls these routines in the first
execution). plpgsql itself is a compiler only for plpgsql statements.
Expressions like c = a + b are processed by SQL parser, which uses
callbacks to plpgsql. And for fixing this issue I need to force SQL parser
to call plpgsql_check callback. The dirty part is fact so plpgsql's
callbacks are partially available, but are marked "static" so I cannot wrap
it directly.

**Comment by okbob** [comment]
should be fixed by 0dee07c

**Comment by preethint** [comment]
Thanks
